# sbsw2019-analysis
R scripts for South Bay Salt Works 2019 microbial diversity analysis. Results published in Klempay _et al._ (2021). https://doi.org/10.1111/1462-2920.15440

All required input files are included in this repository:
- `sbsw2019_16S_preprocessing` takes as input 16S amplicon sequence variant (ASV) abundance matrices generated using [paprica](https://github.com/bowmanjeffs/paprica) v0.5.2, as well as site metadata.
- `sbsw2019_16S_analysis` takes as input intermediate analysis files generated by `sbsw2019_16S_preprocessing`, as well as taxonomic affiliations of ASVs (classified using [paprica](https://github.com/bowmanjeffs/paprica) v0.5.2), and community phylogenetic diversity (Faith's PD) calculated using [guppy](https://matsen.github.io/pplacer/generated_rst/guppy.html) v1.1.alpha19.

Raw amplicon sequence reads and accompanying metadata are available on the NCBI Sequence Read Archive (SRA) under [BioProject PRJNA680352](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA680352/).

## Methods:

Paired-end Illumina reads (MiSeq 2 x 151) were quality-filtered, denoised, and merged using [dada2](https://github.com/benjjneb/dada2) v1.14. The merged reads (fasta files) were then used as input for [paprica](https://github.com/bowmanjeffs/paprica) v0.5.2. Abundant, ambiguous ASVs that could not be classified with paprica were assigned taxonomic data using the web-based [RDP Classifier](https://rdp.cme.msu.edu/classifier) tool. See `taxon_map_modified.csv` files in `paprica_output/`. In order to estimate alpha diversity indices (Chao1 and Faith's PD), 16S reads were first subsampled (n = 12500) to normalize for variation in sequencing depth between samples, and then reanalyzed with dada2 and paprica. See `paprica_output/subsampled/`. Sample ‘5.3’ was excluded due to the low number of total reads (n = 4008). Using paprica output, phylogenetic diversity (Faith's PD) was calculated with [guppy](https://matsen.github.io/pplacer/generated_rst/guppy.html) v1.1.alpha19.

## Please cite:
Klempay, B., Arandia‐Gorostidi, N., Dekas, A. E., Bartlett, D. H., Carr, C. E., Doran, P. T., Dutta, A., Erazo, N., Fisher, L. A., Glass, J. B., Pontefract, A., Som, S. M., Wilson, J. M., Schmidt, B. E., & Bowman, J. S. (2021). Microbial diversity and activity in Southern California salterns and bitterns: analogues for remnant ocean worlds. _Environmental Microbiology_, _23_(7), 3825-3839.
